[Bug]: Component initialization functions panic on invalid or empty configurations

### Bug Description

In Go, it is a well-established architectural anti-pattern for library or component initialization functions to use panic(). A robust node-level daemon like Kepler should gracefully handle misconfigurations or dynamic hardware discovery failures by returning an error from constructors, allowing the caller to decide whether to fail the program, retry, or fallback.

Currently, there are two key components in Kepler that panic during initialization:

NewRedfishCollector (in platform_collector.go) panics if the RedfishDataProvider is nil.
NewAggregatedZone (in energy_zone.go) panics if the dynamically discovered zones list is empty.
If an unexpected hardware layout triggers these panics, the entire Kepler exporter crashes and enters a CrashLoopBackOff, dropping metrics for all other functional subsystems (e.g., RAPL, GPU) on the node.



### Steps to Reproduce

Attempt to initialize an AggregatedZone on a node where the sysfs discovery returns an empty list.
The Kepler daemon will panic with: panic: NewAggregatedZone: zones cannot be empty and crash.
Proposed Solution
I will submit a PR that:

### Expected Behavior

NewRedfishCollector should return (*PlatformCollector, error).
NewAggregatedZone should return (*AggregatedZone, error).
Callers (like prometheus.go and hwmon_power_meter.go) should handle these errors gracefully, log a warning, and continue collecting metrics from the available subsystems.
Actual Behavior
The daemon panics and crashes entirely.

### Environment

- OS: All OS / Linux
- Kubernetes Version: All supported versions
- Container Runtime: All runtimes
- Hardware: All hardware (Issue occurs when dynamic hardware discovery like sysfs or Redfish returns empty lists/nil)
- Deployment Method: Kubernetes DaemonSet


### Logs and Error Messages

```shell
If a user deploys Kepler on a node where sysfs energy zones are empty or the Redfish data provider fails to initialize, the daemon crashes entirely with the following panic signature, rather than logging the error and continuing with the remaining valid collectors:

panic: NewAggregatedZone: zones cannot be empty
goroutine 1 [running]:
github.com/sustainable-computing-io/kepler/internal/device.NewAggregatedZone(...)
        /workspace/internal/device/energy_zone.go:50

(or similarly for RedfishDataProvider cannot be nil in platform_collector.go)
```

### Additional Context

I found this architectural bug while reviewing the codebase. In Go, it is considered a severe anti-pattern for component initialization functions to use `panic()` on unexpected input. A robust node-level daemon like Kepler should gracefully handle misconfigurations or dynamic hardware discovery failures by returning an `error` from constructors, allowing the caller to decide whether to fail the program, retry, or fallback.

I am already working on a PR to fix this! The PR will convert the panics in `NewRedfishCollector` and `NewAggregatedZone` into graceful `error` returns, and update the collectors to log the error and safely continue.

--- a/hack/gen-metric-docs/main.go
+++ b/hack/gen-metric-docs/main.go
@@ -381,7 +381,7 @@ func main() {
 		bmcID:    "test-bmc",
 	}
 	fmt.Println("Creating platform collector...")
-	platformCollector := collector.NewRedfishCollector(mockRedfish, logger)
+	platformCollector, _ := collector.NewRedfishCollector(mockRedfish, logger)
 	fmt.Println("Created platform collector")
 
 	fmt.Println("Extracting metrics from platform collector...")

--- a/internal/device/energy_zone.go
+++ b/internal/device/energy_zone.go
@@ -4,6 +4,7 @@
 package device
 
 import (
+	"errors"
 	"fmt"
 	"math"
 	"sync"
@@ -43,11 +44,10 @@ type AggregatedZone struct {
 
 // NewAggregatedZone creates a new AggregatedZone for zones of the same type
 // The name is taken from the first zone
-// Panics if zones is empty or nil
-func NewAggregatedZone(zones []EnergyZone) *AggregatedZone {
-	// Panic on invalid inputs
+// Returns an error if zones is empty or nil
+func NewAggregatedZone(zones []EnergyZone) (*AggregatedZone, error) {
 	if len(zones) == 0 {
-		panic("NewAggregatedZone: zones cannot be empty")
+		return nil, errors.New("NewAggregatedZone: zones cannot be empty")
 	}
 
 	// Use the first zone's name as the aggregated zone name
@@ -73,7 +73,7 @@ func NewAggregatedZone(zones []EnergyZone) *AggregatedZone {
 		lastReadings:  make(map[zoneKey]Energy),
 		currentEnergy: 0,
 		maxEnergy:     totalMax, // Cache the combined MaxEnergy
-	}
+	}, nil
 }
 
 // Name returns the zone name

--- a/internal/device/energy_zone_test.go
+++ b/internal/device/energy_zone_test.go
@@ -69,7 +69,8 @@ func TestNewAggregatedZone(t *testing.T) {
 		&mockEnergyZone{name: "package", index: 1},
 	}
 
-	az := NewAggregatedZone(zones)
+	az, err := NewAggregatedZone(zones)
+	require.NoError(t, err)
 
 	assert.Equal(t, "package", az.Name())
 	assert.Equal(t, -1, az.Index())
@@ -77,10 +78,9 @@ func TestNewAggregatedZone(t *testing.T) {
 	assert.Len(t, az.zones, 2)
 	assert.NotNil(t, az.lastReadings)
 
-	// Test panic on empty zones
-	assert.Panics(t, func() {
-		NewAggregatedZone([]EnergyZone{})
-	})
+	// Test error on empty zones
+	_, err = NewAggregatedZone([]EnergyZone{})
+	require.Error(t, err)
 }
 
 // TestAggregatedZone_EnergyAggregation tests core energy aggregation functionality
@@ -90,9 +90,8 @@ func TestAggregatedZone_EnergyAggregation(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy, err := az.Energy()
 		require.NoError(t, err)
 		assert.Equal(t, Energy(300), energy)          // 100 + 200
@@ -103,9 +102,8 @@ func TestAggregatedZone_EnergyAggregation(t *testing.T) {
 		zones := []EnergyZone{
 			&mockEnergyZone{name: "package", index: 0, energy: 150, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy, err := az.Energy()
 		require.NoError(t, err)
 		assert.Equal(t, Energy(150), energy)
@@ -115,9 +113,8 @@ func TestAggregatedZone_EnergyAggregation(t *testing.T) {
 		// Test that first reading doesn't double-count
 		zone := &mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// First reading should return current energy
 		energy1, err := az.Energy()
 		require.NoError(t, err)
@@ -142,9 +139,8 @@ func TestAggregatedZone_WrapHandling(t *testing.T) {
 		zone0 := &mockEnergyZone{name: "package", index: 0, energy: 900, maxEnergy: 1000}
 		zone1 := &mockEnergyZone{name: "package", index: 1, energy: 800, maxEnergy: 1000}
 		zones := []EnergyZone{zone0, zone1}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// First reading: 900 + 800 = 1700
 		energy1, err := az.Energy()
 		require.NoError(t, err)
@@ -162,9 +158,8 @@ func TestAggregatedZone_WrapHandling(t *testing.T) {
 	t.Run("MultipleWraps", func(t *testing.T) {
 		zone := &mockEnergyZone{name: "package", index: 0, energy: 900, maxEnergy: 1000}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// First reading: 900
 		energy1, err := az.Energy()
 		require.NoError(t, err)
@@ -189,9 +184,8 @@ func TestAggregatedZone_WrapHandling(t *testing.T) {
 		// Test handling of readings that go backward (faulty hardware)
 		zone := &mockEnergyZone{name: "package", index: 0, energy: 500, maxEnergy: 1000}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy1, err := az.Energy()
 		require.NoError(t, err)
 		assert.Equal(t, Energy(500), energy1)
@@ -211,8 +205,8 @@ func TestAggregatedZone_WrapHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 0},
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: 0},
 		}
-
-		az := NewAggregatedZone(zones)
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		assert.Equal(t, Energy(0), az.MaxEnergy())
 
 		// Should not panic with zero MaxEnergy
@@ -229,9 +223,8 @@ func TestAggregatedZone_ErrorHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000, err: fmt.Errorf("read error")},
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy, err := az.Energy()
 		assert.Error(t, err)
 		assert.Contains(t, err.Error(), "no valid energy readings")
@@ -243,9 +236,8 @@ func TestAggregatedZone_ErrorHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000, err: fmt.Errorf("error1")},
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: 1000, err: fmt.Errorf("error2")},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy, err := az.Energy()
 		assert.Error(t, err)
 		assert.Contains(t, err.Error(), "no valid energy readings")
@@ -260,8 +252,8 @@ func TestAggregatedZone_MaxEnergyHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		assert.Equal(t, Energy(2000), az.MaxEnergy())
 	})
 
@@ -271,8 +263,8 @@ func TestAggregatedZone_MaxEnergyHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 2, energy: 300, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		expectedMaxEnergy := Energy(3000)
 
 		// Multiple calls should return the same cached value
@@ -281,7 +273,7 @@ func TestAggregatedZone_MaxEnergyHandling(t *testing.T) {
 		assert.Equal(t, expectedMaxEnergy, az.MaxEnergy())
 
 		// Verify cached value is used after Energy() calls
-		_, err := az.Energy()
+		_, err = az.Energy()
 		require.NoError(t, err)
 		assert.Equal(t, expectedMaxEnergy, az.MaxEnergy())
 	})
@@ -294,9 +286,8 @@ func TestAggregatedZone_MaxEnergyHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: largeMaxEnergy},
 			&mockEnergyZone{name: "package", index: 2, energy: 300, maxEnergy: largeMaxEnergy},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Should not panic and should handle overflow gracefully
 		maxEnergy := az.MaxEnergy()
 		t.Logf("MaxEnergy with potential overflow: %d", maxEnergy)
@@ -328,9 +319,8 @@ func TestAggregatedZone_LargeValueHandling(t *testing.T) {
 			maxEnergy: maxEnergy,
 		}
 		zones := []EnergyZone{zone0, zone1}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Should handle large values without overflow
 		energy1, err := az.Energy()
 		require.NoError(t, err)
@@ -355,9 +345,8 @@ func TestAggregatedZone_LargeValueHandling(t *testing.T) {
 			maxEnergy: maxEnergy,
 		}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy1, err := az.Energy()
 		require.NoError(t, err)
 		assert.Equal(t, Energy(maxEnergy-100), energy1)
@@ -382,9 +371,8 @@ func TestAggregatedZone_LargeValueHandling(t *testing.T) {
 			maxEnergy: largeMaxEnergy,
 		}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy1, err := az.Energy()
 		require.NoError(t, err)
 		assert.Equal(t, Energy(1000), energy1)
@@ -413,9 +401,8 @@ func TestAggregatedZone_ConcurrentAccess(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Test concurrent reads don't cause race conditions
 		done := make(chan bool, 10)
 		for i := 0; i < 10; i++ {
@@ -434,11 +421,10 @@ func TestAggregatedZone_ConcurrentAccess(t *testing.T) {
 	t.Run("ConcurrentReadsWithUpdates", func(t *testing.T) {
 		zone := &mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Initialize state
-		_, err := az.Energy()
+		_, err = az.Energy()
 		require.NoError(t, err)
 
 		// Concurrent access with zone energy updates
@@ -493,11 +479,10 @@ func TestAggregatedZone_StateManagement(t *testing.T) {
 	t.Run("BasicStateTracking", func(t *testing.T) {
 		zone := &mockEnergyZone{name: "package", index: 0, energy: 500, maxEnergy: 1000}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// First reading should initialize last reading
-		_, err := az.Energy()
+		_, err = az.Energy()
 		require.NoError(t, err)
 
 		zoneID := zoneKey{"package", 0}
@@ -516,11 +501,10 @@ func TestAggregatedZone_StateManagement(t *testing.T) {
 	t.Run("ZoneIDGeneration", func(t *testing.T) {
 		zone := &mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// First call should create last reading entry
-		_, err := az.Energy()
+		_, err = az.Energy()
 		require.NoError(t, err)
 		assert.Len(t, az.lastReadings, 1)
 
@@ -535,9 +519,8 @@ func TestAggregatedZone_StateManagement(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 100, maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, energy: 200, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Concurrent first-time access
 		done := make(chan bool, 10)
 		for i := 0; i < 10; i++ {
@@ -576,9 +559,8 @@ func TestAggregatedZone_EdgeCases(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 900, maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, energy: 900, maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		energy1, err := az.Energy()
 		require.NoError(t, err)
 		assert.Equal(t, Energy(1800), energy1)
@@ -612,9 +594,8 @@ func TestAggregatedZone_EdgeCases(t *testing.T) {
 			maxEnergy: largeMax,
 		}
 		zones := []EnergyZone{zone0, zone1}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Should work without overflow
 		energy1, err := az.Energy()
 		require.NoError(t, err)
@@ -640,9 +621,8 @@ func TestAggregatedZone_Power(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(50.0), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(75.5), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.Equal(t, Power(125.5), power) // 50.0 + 75.5
@@ -652,9 +632,8 @@ func TestAggregatedZone_Power(t *testing.T) {
 		zones := []EnergyZone{
 			&mockEnergyZone{name: "package", index: 0, power: Power(42.5), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.Equal(t, Power(42.5), power)
@@ -665,9 +644,8 @@ func TestAggregatedZone_Power(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(0.0), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(0.0), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.Equal(t, Power(0.0), power)
@@ -680,9 +658,8 @@ func TestAggregatedZone_Power(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 2, power: Power(45.25), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 3, power: Power(60.0), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.InDelta(t, 160.75, float64(power), 0.01) // 25.0 + 30.5 + 45.25 + 60.0
@@ -693,9 +670,8 @@ func TestAggregatedZone_Power(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(12.345), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(67.890), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.InDelta(t, 80.235, float64(power), 0.001)
@@ -709,9 +685,8 @@ func TestAggregatedZone_PowerErrorHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(50.0), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(75.5), powerErr: fmt.Errorf("sensor failure"), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		assert.Error(t, err)
 		assert.Contains(t, err.Error(), "failed to read power from zone")
@@ -724,9 +699,8 @@ func TestAggregatedZone_PowerErrorHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(0.0), powerErr: fmt.Errorf("hw error"), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(75.5), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		assert.Error(t, err)
 		assert.Contains(t, err.Error(), "failed to read power from zone")
@@ -738,9 +712,8 @@ func TestAggregatedZone_PowerErrorHandling(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, powerErr: fmt.Errorf("error1"), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, powerErr: fmt.Errorf("error2"), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		assert.Error(t, err)
 		assert.Contains(t, err.Error(), "failed to read power from zone")
@@ -755,9 +728,8 @@ func TestAggregatedZone_PowerConcurrentAccess(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(50.0), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(75.5), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		done := make(chan bool, 20)
 		results := make(chan Power, 20)
 
@@ -794,9 +766,8 @@ func TestAggregatedZone_PowerConcurrentAccess(t *testing.T) {
 		zone0 := &mockEnergyZone{name: "package", index: 0, power: 50.0, maxEnergy: 1000}
 		zone1 := &mockEnergyZone{name: "package", index: 1, power: 75.5, maxEnergy: 1000}
 		zones := []EnergyZone{zone0, zone1}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		done := make(chan bool, 30)
 		results := make(chan Power, 20)
 
@@ -846,9 +817,8 @@ func TestAggregatedZone_PowerAndEnergyIndependence(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, energy: 100, power: Power(50.0), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, energy: 200, power: Power(75.5), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Call Power() first
 		power1, err := az.Power()
 		require.NoError(t, err)
@@ -873,9 +843,8 @@ func TestAggregatedZone_PowerAndEnergyIndependence(t *testing.T) {
 	t.Run("EnergyDoesNotAffectPower", func(t *testing.T) {
 		zone := &mockEnergyZone{name: "package", index: 0, energy: 100, power: Power(50.0), maxEnergy: 1000}
 		zones := []EnergyZone{zone}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		// Call Energy() first
 		energy1, err := az.Energy()
 		require.NoError(t, err)
@@ -906,9 +875,8 @@ func TestAggregatedZone_PowerEdgeCases(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(999999.99), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(888888.88), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.InDelta(t, 1888888.87, float64(power), 0.01)
@@ -919,9 +887,8 @@ func TestAggregatedZone_PowerEdgeCases(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 0, power: Power(0.0001), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 1, power: Power(0.0002), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.InDelta(t, 0.0003, float64(power), 0.00001)
@@ -933,9 +900,8 @@ func TestAggregatedZone_PowerEdgeCases(t *testing.T) {
 			&mockEnergyZone{name: "package", index: 1, power: Power(0.0), maxEnergy: 1000},
 			&mockEnergyZone{name: "package", index: 2, power: Power(50.0), maxEnergy: 1000},
 		}
-
-		az := NewAggregatedZone(zones)
-
+		az, err := NewAggregatedZone(zones)
+		require.NoError(t, err)
 		power, err := az.Power()
 		require.NoError(t, err)
 		assert.Equal(t, Power(150.0), power)

--- a/internal/device/hwmon_power_meter.go
+++ b/internal/device/hwmon_power_meter.go
@@ -213,7 +213,7 @@ func (h *hwmonPowerMeter) groupZonesByName(zones []EnergyZone) []EnergyZone {
 		// LIMITATION: aggregation occurs when the devices are different with coincidentally
 		// the same labels. This should not happen. Ideally, Kepler identifies whether the zones with same
 		// name occur due to multi-socket CPU or independent devices.
-		aggregated := NewAggregatedZone(zones)
+		aggregated, _ := NewAggregatedZone(zones)
 		result = append(result, aggregated)
 		h.logger.Debug("Created aggregated zone",
 			"name", name,

--- a/internal/device/rapl_sysfs_power_meter.go
+++ b/internal/device/rapl_sysfs_power_meter.go
@@ -158,26 +158,23 @@ func (r *raplPowerMeter) groupZonesByName(stdZoneMap map[zoneKey]EnergyZone) []E
 	// Group zones by base name (e.g., "package", "dram")
 	zoneGroups := make(map[string][]EnergyZone)
 
-	for key, zone := range stdZoneMap {
-		zoneGroups[key.name] = append(zoneGroups[key.name], zone)
+	for _, zone := range stdZoneMap {
+		name := zone.Name()
+		zoneGroups[name] = append(zoneGroups[name], zone)
 	}
 
-	// Create aggregated zones for duplicates, keep single zones as-is
 	var result []EnergyZone
 	for name, zones := range zoneGroups {
 		if len(zones) == 1 {
-			// Single zone - use as-is
 			result = append(result, zones[0])
 			continue
-
 		}
 
 		// Multiple zones with same name - create AggregatedZone
-		aggregated := NewAggregatedZone(zones)
+		aggregated, _ := NewAggregatedZone(zones)
 		result = append(result, aggregated)
 		r.logger.Debug("Created aggregated zone",
 			"name", name,
-			"zone_count", len(zones),
 			"zones", r.zoneNames(zones))
 	}
 

--- a/internal/exporter/prometheus/collector/platform_collector.go
+++ b/internal/exporter/prometheus/collector/platform_collector.go
@@ -4,6 +4,7 @@
 package collector
 
 import (
+	"errors"
 	"log/slog"
 	"time"
 
@@ -38,9 +39,9 @@ type PlatformCollector struct {
 }
 
 // NewRedfishCollector creates a new platform collector
-func NewRedfishCollector(redfish RedfishDataProvider, logger *slog.Logger) *PlatformCollector {
+func NewRedfishCollector(redfish RedfishDataProvider, logger *slog.Logger) (*PlatformCollector, error) {
 	if redfish == nil {
-		panic("RedfishDataProvider cannot be nil - platform collector requires a data provider to function")
+		return nil, errors.New("RedfishDataProvider cannot be nil - platform collector requires a data provider to function")
 	}
 	if logger == nil {
 		logger = slog.Default()
@@ -57,7 +58,7 @@ func NewRedfishCollector(redfish RedfishDataProvider, logger *slog.Logger) *Plat
 			[]string{"source", "node_name", "bmc_id", "chassis_id", "source_id", "source_name", "source_type"},
 			nil,
 		),
-	}
+	}, nil
 }
 
 // Describe sends the descriptors of platform metrics to the provided channel

--- a/internal/exporter/prometheus/collector/platform_collector_test.go
+++ b/internal/exporter/prometheus/collector/platform_collector_test.go
@@ -92,7 +92,8 @@ func TestNewRedfishCollector(t *testing.T) {
 		bmcID:    "test-bmc",
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	require.NotNil(t, collector)
 	assert.Equal(t, "test-node", collector.nodeName)
@@ -102,13 +103,14 @@ func TestNewRedfishCollector(t *testing.T) {
 	assert.Equal(t, mockProvider, collector.redfish)
 }
 
-func TestNewRedfishCollector_ValidationPanics(t *testing.T) {
+func TestNewRedfishCollector_ValidationError(t *testing.T) {
 	logger := slog.New(slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelError}))
 
-	t.Run("Nil data provider panics", func(t *testing.T) {
-		assert.Panics(t, func() {
-			NewRedfishCollector(nil, logger)
-		}, "Should panic when RedfishDataProvider is nil")
+	t.Run("Nil data provider returns error", func(t *testing.T) {
+		collector, err := NewRedfishCollector(nil, logger)
+		require.Error(t, err)
+		assert.Nil(t, collector)
+		assert.Contains(t, err.Error(), "RedfishDataProvider cannot be nil")
 	})
 
 	t.Run("Nil logger uses default", func(t *testing.T) {
@@ -117,7 +119,8 @@ func TestNewRedfishCollector_ValidationPanics(t *testing.T) {
 			bmcID:    "test-bmc",
 		}
 
-		collector := NewRedfishCollector(mockProvider, nil)
+		collector, err := NewRedfishCollector(mockProvider, nil)
+		require.NoError(t, err)
 		require.NotNil(t, collector)
 		assert.NotNil(t, collector.logger, "Should use default logger when nil is passed")
 	})
@@ -131,7 +134,8 @@ func TestPlatformCollector_Describe(t *testing.T) {
 		bmcID:    "test-bmc",
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	// Create a channel to collect descriptors
 	ch := make(chan *prometheus.Desc, 10)
@@ -199,7 +203,8 @@ func TestPlatformCollector_Collect_Success(t *testing.T) {
 		powerReading: powerReading,
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	// Create registry and register collector
 	registry := prometheus.NewRegistry()
@@ -261,7 +266,8 @@ func TestPlatformCollector_Collect_Error(t *testing.T) {
 		err:      errors.New("BMC connection failed"),
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	// Create registry and register collector
 	registry := prometheus.NewRegistry()
@@ -284,7 +290,8 @@ func TestPlatformCollector_Collect_NilReading(t *testing.T) {
 		powerReading: nil, // nil reading
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	// Create registry and register collector
 	registry := prometheus.NewRegistry()
@@ -313,7 +320,8 @@ func TestPlatformCollector_Collect_EmptyReadings(t *testing.T) {
 		powerReading: powerReading,
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	// Create registry and register collector
 	registry := prometheus.NewRegistry()
@@ -353,7 +361,8 @@ func TestPlatformCollector_Collect_SingleChassis(t *testing.T) {
 		powerReading: powerReading,
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	// Create registry and register collector
 	registry := prometheus.NewRegistry()
@@ -404,7 +413,8 @@ func TestPlatformCollector_Collect_ParallelCollection(t *testing.T) {
 		powerReading: powerReading,
 	}
 
-	collector := NewRedfishCollector(mockProvider, logger)
+	collector, err := NewRedfishCollector(mockProvider, logger)
+	require.NoError(t, err)
 
 	// Create registry and register collector
 	registry := prometheus.NewRegistry()
@@ -494,7 +504,8 @@ func TestPlatformCollector_Collect_MetricLabelsValidation(t *testing.T) {
 				powerReading: powerReading,
 			}
 
-			collector := NewRedfishCollector(mockProvider, logger)
+			collector, err := NewRedfishCollector(mockProvider, logger)
+			require.NoError(t, err)
 
 			// Create registry and register collector
 			registry := prometheus.NewRegistry()

--- a/internal/exporter/prometheus/prometheus.go
+++ b/internal/exporter/prometheus/prometheus.go
@@ -166,8 +166,9 @@ func CreateCollectors(pm Monitor, applyOpts ...OptionFn) (map[string]prom.Collec
 	collectors["gpu_info"] = collector.NewGPUInfoCollector(pm, opts.nodeName)
 
 	// Add platform collector if platform data provider is available
-	if opts.platformDataProvider != nil {
-		collectors["platform"] = collector.NewRedfishCollector(opts.platformDataProvider, opts.logger)
+	platformCollector, err := collector.NewRedfishCollector(opts.platformDataProvider, opts.logger)
+	if err == nil {
+		collectors["platform"] = platformCollector
 	}
 
 	return collectors, nil

--- a/internal/exporter/prometheus/prometheus_test.go
+++ b/internal/exporter/prometheus/prometheus_test.go
@@ -14,7 +14,9 @@ import (
 	prom "github.com/prometheus/client_golang/prometheus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
+	"github.com/sustainable-computing-io/kepler/config"
 	"github.com/sustainable-computing-io/kepler/internal/monitor"
+	"github.com/sustainable-computing-io/kepler/internal/platform/redfish"
 )
 
 // MockMonitor mocks the Monitor interface
@@ -280,6 +282,19 @@ func TestWithOptions(t *testing.T) {
 		assert.True(t, opts.debugCollectors["process"])
 		assert.True(t, opts.debugCollectors["custom"])
 	})
+
+	t.Run("WithMetricsLevel", func(t *testing.T) {
+		opts := DefaultOpts()
+		WithMetricsLevel(1)(&opts)
+		assert.Equal(t, config.Level(1), opts.metricsLevel)
+	})
+
+	t.Run("WithPlatformDataProvider", func(t *testing.T) {
+		opts := DefaultOpts()
+		// Any mock struct can be used, nil is fine for this interface check
+		WithPlatformDataProvider(nil)(&opts)
+		assert.Nil(t, opts.platformDataProvider)
+	})
 }
 
 func TestDefaultOpts(t *testing.T) {
@@ -314,6 +329,12 @@ func TestExporter_Integration(t *testing.T) {
 	mockMonitor.AssertExpectations(t)
 }
 
+type mockPlatformProvider struct{}
+
+func (m *mockPlatformProvider) Power() (*redfish.PowerReading, error) { return nil, nil }
+func (m *mockPlatformProvider) NodeName() string                      { return "test-node" }
+func (m *mockPlatformProvider) BMCID() string                         { return "test-bmc" }
+
 func TestExporter_CreateCollectors(t *testing.T) {
 	mockMonitor := &MockMonitor{}
 	mockMonitor.On("DataChannel").Return(make(<-chan struct{}))
@@ -323,6 +344,7 @@ func TestExporter_CreateCollectors(t *testing.T) {
 		mockMonitor,
 		WithLogger(slog.Default()),
 		WithProcFSPath("/proc"),
+		WithPlatformDataProvider(&mockPlatformProvider{}),
 	)
 	time.Sleep(50 * time.Millisecond)
 
@@ -330,5 +352,5 @@ func TestExporter_CreateCollectors(t *testing.T) {
 	mockMonitor.AssertExpectations(t)
 
 	assert.NoError(t, err)
-	assert.Len(t, coll, 4) // build_info, power, cpu_info, gpu_info
+	assert.Len(t, coll, 5) // build_info, power, cpu_info, gpu_info, platform
 }

--- a/internal/exporter/stdout/stdout_test.go
+++ b/internal/exporter/stdout/stdout_test.go
@@ -113,7 +113,7 @@ func TestExporter_InitRunShotdown(t *testing.T) {
 		mockMonitor := &MockMonitor{}
 		mockMonitor.On("Snapshot").Return(&monitor.Snapshot{Node: getTestNodeData()}, nil)
 		out := &dummyWriteCloser{&bytes.Buffer{}}
-		exporter := NewExporter(mockMonitor, WithOutput(out), WithInterval(1*time.Second))
+		exporter := NewExporter(mockMonitor, WithOutput(out), WithInterval(100*time.Millisecond))
 		err := exporter.Init()
 		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
 		go func() {

